### PR TITLE
Add reads2graph

### PR DESCRIPTION
**Pull request by Jappy0**
reads2graph is a practical method for constructing an edit-distance-based read graph for a short-read sequencing dataset via minimizer and order min hash bucketing and graph traversal.

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by Jappy0** [comment]
@BiocondaBot please add label

**Comment by Jappy0** [comment]
Hi [martin-g](https://github.com/martin-g),

Re: https://github.com/bioconda/bioconda-recipes/pull/48131#issuecomment-2139142593. You are right—the easiest way to create a new recipe. Thanks so much.

By creating a new recipe, I found that I got stuck on bioconda-utils testing because I used `git add .` to add my recipe to git, and I didn't delete my previous published recipe locally and its branch. 

Anyway, all the tests passed for this recipe. Thanks again. 

Best,
Jappy

File: recipes/reads2graph/build.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+set -ex
+
+# Add the Conda environment's bin directory to the PATH
+export PATH="${PREFIX}/bin:${PATH}"
+
+#Set the CXX environment variable to the specific g++ version
+export CXX=$(which g++)
+export CC=$(which gcc)
```
Comment (martin-g) [inline recipes/reads2graph/build.sh:9]: These two lines are not needed

File: recipes/reads2graph/build.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+# Check if build directory exists and delete it if it does
+if [ -d "build" ]; then
```
Comment (martin-g) [inline recipes/reads2graph/build.sh:12]: In CI run this folder should not exist

File: recipes/reads2graph/build.sh [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+cmake -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_INSTALL_PREFIX=${PREFIX} \
+      -DCMAKE_MAKE_PROGRAM=make \
+      -DCMAKE_CXX_COMPILER=${CXX} \
+      -DCMAKE_TOOLCHAIN_FILE="${RECIPE_DIR}/cross-linux.cmake" \
+      -DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS} -std=gnu++20 -fopenmp -static-libgcc -static-libstdc++ -isystem ${PREFIX}/include -isystem ${PREFIX}/include/seqan3/submodules/sdsl-lite/include -isystem ${PREFIX}/include/seqan3/submodules/cereal/include -fconcepts -g -Wall -lyaml-cpp -Wno-interference-size -Wno-maybe-uninitialized" \
+      ..
+
+make clean
+make
```
Comment (martin-g) [inline recipes/reads2graph/build.sh:33]:
```suggestion
make -j ${CPU_COUNT}
```

File: recipes/reads2graph/build.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+# Configure the build
+cmake -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_INSTALL_PREFIX=${PREFIX} \
+      -DCMAKE_MAKE_PROGRAM=make \
+      -DCMAKE_CXX_COMPILER=${CXX} \
+      -DCMAKE_TOOLCHAIN_FILE="${RECIPE_DIR}/cross-linux.cmake" \
+      -DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS} -std=gnu++20 -fopenmp -static-libgcc -static-libstdc++ -isystem ${PREFIX}/include -isystem ${PREFIX}/include/seqan3/submodules/sdsl-lite/include -isystem ${PREFIX}/include/seqan3/submodules/cereal/include -fconcepts -g -Wall -lyaml-cpp -Wno-interference-size -Wno-maybe-uninitialized" \
+      ..
+
+make clean
```
Comment (martin-g) [inline recipes/reads2graph/build.sh:32]: Not really needed, but it does not do any harm too!

File: recipes/reads2graph/cross-linux.cmake [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# this one is important
```
Comment (martin-g) [inline recipes/reads2graph/cross-linux.cmake:1]: Why do you need this file ?
AFAIS linux x86_64 is the only build you do (osx Intel is disabled, and linux aarch64 is not enabled in meta.yaml)

File: recipes/reads2graph/build.sh (hunk 1 above)
Comment (Jappy0) [reply recipes/reads2graph/build.sh:9]: Yes, these two lines are indeed not needed. I used this to help me debug. I will remove it. Thanks, Martin.

File: recipes/reads2graph/build.sh (hunk 2 above)
Comment (Jappy0) [reply recipes/reads2graph/build.sh:12]: ditto

File: recipes/reads2graph/build.sh (hunk 3 above)
Comment (Jappy0) [reply recipes/reads2graph/build.sh:33]: `make -j ${CPU_COUNT}` should work, but I got stuck during compiling with using it. Let me try again.

File: recipes/reads2graph/build.sh (hunk 4 above)
Comment (Jappy0) [reply recipes/reads2graph/build.sh:32]: ditto, I will remove it.

File: recipes/reads2graph/build.sh (hunk 3 above)
Comment (Jappy0) [reply recipes/reads2graph/build.sh:20]: By trying to use `make -j ${CPU_COUNT}` again, I found it got stuck at 37% of the building process, which is the same as my previous testing. I have no idea why this did not work. With one CPU core, it works smoothly and fast in both local compiling and CI. So let's remove `-j ${CPU_COUNT}`. Is this ok?

File: recipes/reads2graph/cross-linux.cmake (hunk 5 above)
Comment (Jappy0) [reply recipes/reads2graph/cross-linux.cmake:1]: Without `cross-linux.cmake`, CMake cannot work during the test process in Linux. CI got stopped for testing and reported an error log missing. After trying many times, I resolved it by adding this file following the tips (copied below) on [Anaconda Comioler-tools](https://conda.io/projects/conda-build/en/latest/resources/compiler-tools.html#compiler-tools).

**An aside on CMake and sysroots**
Anaconda's compilers for Linux are built with something called crosstool-ng. They include not only GCC, but also a "sysroot" with glibc, as well as the rest of the toolchain (binutils). Ordinarily, the sysroot is something that your system provides, and it is what establishes the libc compatibility bound for your compiled code. Any compilation that uses a sysroot other than the system sysroot is said to be "cross-compiling." When the target OS and the build OS are the same, it is called a "pseudo-cross-compiler." This is the case for normal builds with Anaconda's compilers on Linux.

Unfortunately, some software tools do not handle sysroots in intuitive ways. CMake is especially bad for this. Even though the compiler itself understands its own sysroot, CMake insists on ignoring that. We've filed issues at:

https://gitlab.kitware.com/cmake/cmake/issues/17483

Additionally, this Stack Overflow issue has some more information: https://stackoverflow.com/questions/36195791/cmake-missing-sysroot-when-cross-compiling

In order to teach CMake about the sysroot, you must do additional work. As an example, please see our recipe for libnetcdf at [AnacondaRecipes/libnetcdf-feedstock](https://github.com/AnacondaRecipes/libnetcdf-feedstock/tree/master/recipe)

In particular, you'll need to copy the cross-linux.cmake file there, and reference it in your build.sh file:

CMAKE_PLATFORM_FLAGS+=(-DCMAKE_TOOLCHAIN_FILE="${RECIPE_DIR}/cross-linux.cmake")

cmake -DCMAKE_INSTALL_PREFIX=${PREFIX} \
  ${CMAKE_PLATFORM_FLAGS[@]} \
  ${SRC_DIR}

Comment (Jappy0) [reply recipes/reads2graph/cross-linux.cmake:1]: Btw, osx is disabled because one dependency of reads2graph is not available in osx. 

is `- {{ compiler('c') }}` meaning that linux aarch64 enabled? If it is, the test on Linux aarch64 should have passed.

Comment (martin-g) [reply recipes/reads2graph/cross-linux.cmake:1]: To enable `linux-aarch64` you need do add `extra > additional-platforms > - linux-aarch64`, e.g. https://github.com/bioconda/bioconda-recipes/blob/d77ba3a86e386795e9eddaa336aba788117cf635/recipes/3seq/meta.yaml#L37-L39

**Review by martin-g — APPROVED** [review]

File: recipes/reads2graph/cross-linux.cmake (hunk 5 above)
Comment (Jappy0) [reply recipes/reads2graph/cross-linux.cmake:1]: > extra: 
>    additional-platforms: 
>      - linux-aarch64

Ok, Got it. Thanks so much, let me update it next upgrade.
